packages/drivers/odsp-driver/src/odspDriverUrlResolver2.ts

```diff
@@ -0,0 +1,186 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IRequest } from "@fluidframework/core-interfaces";
+import { IResolvedUrl, IUrlResolver } from "@fluidframework/driver-definitions";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { getLocatorFromOdspUrl, storeLocatorInOdspUrl, encodeOdspFluidDataStoreLocator } from "./odspFluidFileLink";
+import { resolveDataStore } from "./resolveDataStore";
+import { IOdspResolvedUrl, OdspDocumentInfo, OdspFluidDataStoreLocator, SharingLinkHeader } from "./contracts";
+import { createOdspCreateContainerRequest } from "./createOdspCreateContainerRequest";
+import { createOdspUrl } from "./createOdspUrl";
+import { resolveRequest } from "./odspDriverUrlResolver";
+import { getShareLink } from "./graph";
+import { IdentityType, TokenFetchOptions } from "./tokenFetch";
+
+export class OdspDriverUrlResolver2 implements IUrlResolver {
+    private readonly shareLinkMap = new Map<string, Promise<string>>();
+
+    public constructor(
+        private readonly getSharingLinkToken:
+            (options: TokenFetchOptions, isForFileDefaultUrl: boolean) => Promise<string | null>,
+        private readonly identityType: IdentityType = "Enterprise",
+        private readonly logger?: ITelemetryLogger,
+        private readonly appName?: string,
+    ) { }
+
+    public createCreateNewRequest(siteUrl: string, driveId: string, filePath: string, fileName: string) {
+        return createOdspCreateContainerRequest(siteUrl, driveId, filePath, fileName);
+    }
+
+    private getKey(resolvedUrl: IOdspResolvedUrl): string {
+        return `${resolvedUrl.siteUrl},${resolvedUrl.driveId},${resolvedUrl.itemId}`;
+    }
+
+    /**
+     * Takes an already generated data store url (from requestUrl) and appends a path to the
+     * existing data store information.
+     */
+    public appendDataStorePath(requestUrl: URL, pathToAppend: string): string | undefined {
```

Comment:
Mara actually just submitted a PR adding some logic to this method to better handle paths that already have query params etc. We can update the logic in a future iteration though, especially since the name changed.

This will also have to be actually accessible off of component runtimes/contexts like getAbsoluteUrl is today so more work here too. (orrrr you can just expose the resolver to the context :D)

Reply:
Took Mara's change in this api.